Add APRS device tracker component

PhilRW opened the pull request:
## Description:
This new APRS component keeps open a connection to the APRS-IS infrastructure so messages generated by filtered callsigns can be immediately acted upon. Any messages with certain values for the 'format' key are position reports and are parsed into device tracker entities.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9056

## Example entry for `configuration.yaml` (if applicable):
```yaml
device_tracker:
  - platform: aprs
    callsigns:
      - 'XX0FOO*'
      - 'YY0BAR-1'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

Commit:
Add APRS device tracker component

This component keeps open a connection to the APRS-IS infrastructure so
messages generated by filtered callsigns can be immediately acted upon.
Any messages with certain values for the 'format' key are position
reports and are parsed into device tracker entities.

syssi commented:
Yeay! First(?) hamradio component for HA! :-)

syssi commented on homeassistant/components/aprs/device_tracker.py:
```diff
@@ -0,0 +1,129 @@
+            lon = msg[ATTR_LONGITUDE]
+
+            attrs = {}
+            if ATTR_ALTITUDE in msg:
```
You've accessed latitude and longitude in line 114 already. This check isn't needed.

PhilRW replied:
That's 'altitude', not 'latitude'. 🙂 I want to include it in the attributes if it's present.

syssi replied:
Ups! My fault.

syssi replied:
You could clean-up the code here if you introduce a for loop for all handled attributes.

PhilRW replied:
I originally had it that way but split them out because I wanted to change the names. But I removed the name changing so I'll change it back to a loop.

syssi commented on homeassistant/components/aprs/device_tracker.py:
```diff
@@ -0,0 +1,129 @@
+def setup_scanner(hass, config, see, discovery_info=None):
+    """Set up the APRS tracker."""
+    callsigns = config.get(CONF_CALLSIGNS)
+    fltr = make_filter(callsigns)
```
Call it "filter". ;-)

PhilRW replied:
I did originally, but then the linters told me that I was overriding Python's built-in `filter` keyword. 😕

syssi replied:
What about `traffic_filter` or `callsign_filter`?

PhilRW replied:
I think I'll go with `server_filter` since it filters server-side traffic.

syssi commented on homeassistant/components/aprs/device_tracker.py:
```diff
@@ -0,0 +1,129 @@
+        self.host = host
+        self.fltr = fltr
+        self.see = see
+        self.ais = aprslib.IS(self.callsign, host=self.host, port=14580)
```
Please make the port configurable, too.

PhilRW replied:
I'll make it a constant but that port number is required when doing server-side filtering so I don't want it end-user configurable just yet.

syssi replied:
It isn't the port of the tcp connection?

PhilRW replied:
There are different ports for different purposes. The first time I connected, I got the entire feed which I definitely didn't want. See [Connecting to APRS-IS](http://aprs-is.net/Connecting.aspx).

syssi replied:
Okay. 14580 is the default tcp port for the "filter feed". It isn't configurable to avoid confusion.

PhilRW commented:
I'm trying to figure out how to pass the `PlatformNotReady` back to the main thread.

PhilRW commented:
I don't quite get what's going on here with codecov. I've rebased onto the latest dev branch but it says I've changed all these files that I haven't touched in my commit(s).

brent20 commented:
I just wanted to drop by and add some support for this component. I was just thinking about APRS and telemetry and thought "wouldn't it be cool for Home Assistant to be able to receive and transmit telemetry on the APRS-IS network!?" Sure enough, I come to find that the community is already working on this! Thanks! @PhilRW!

PhilRW commented:
Long-term we could send locations and messages to the network, but I'm starting simple with a device tracker component.

PhilRW commented:
Perhaps some additional developer documentation is needed when adding a codeowner.

PhilRW commented:
@syssi Now that I finally have it passing CI, could you take another look? 🙂

balloob commented on homeassistant/components/aprs/device_tracker.py:
```diff
@@ -0,0 +1,186 @@
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, aprs_disconnect)
+
+    if not aprs_listener.start_event.wait(timeout):
+        raise PlatformNotReady("Timeout waiting for APRS to connect.")
```
This won't work for device tracker, because it is using older infrastructure than the `async_setup_platform` code for which the `PlatformNotReady` exception is meant.

PhilRW replied:
I'm ready to clean it up but things may have shifted in the interim -- any suggestions are appreciated.

balloob replied:
Print an error and return. In a future PR we can migrate this to config entries and you can opt-in to the new way the device tracker does things, using entities. It's a bit more complicated and I don't want this PR to blow up but instead get it merged. Got side tracked but still reviewing

balloob commented on tests/components/aprs/test_device_tracker.py:
```diff
@@ -0,0 +1,338 @@
+TEST_PASSWORD = 'testpass'
+
+
+class TestAprsDeviceTracker(unittest.TestCase):
```
We have been moving away from class based tests. Instead, just make it all functions and it should work.

balloob reviewed (approved):
Ok to merge when tests converted from class functions to standalone functions and `PlatformNotReady` replaced with error logging.

Commit:
Log errors and return if startup failure

Commit:
Fix unit tests

PhilRW commented:
Fixed.
